File: src/main/java/io/trydent/olimpo/http/HttpRequest.java
package io.trydent.olimpo.http;

import io.trydent.olimpo.type.Type;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpServerRequest;
import org.slf4j.Logger;

import java.util.Arrays;

import static io.vertx.ext.web.Router.router;
import static org.slf4j.LoggerFactory.getLogger;

public interface HttpRequest extends Type.As<Handler<HttpServerRequest>> {
    static HttpRequest routeSwitch(final Vertx vertx, final HttpRoute... routes) {
      return new HttpRouteSwitch(vertx, Arrays.copyOf(routes, routes.length));
    }

    static HttpRequest single(Vertx vertx, HttpRoute route) {
      return new HttpSingle(vertx, route);
    }
}

final class HttpRouteSwitch implements HttpRequest {
  private static final Logger log = getLogger(HttpRouteSwitch.class);

  private final Vertx vertx;
  private final HttpRoute[] routes;

  HttpRouteSwitch(final Vertx vertx, final HttpRoute[] routes) {
    this.vertx = vertx;
    this.routes = routes;
  }

  @Override
  public final Handler<HttpServerRequest> get() {
    final var router = router(vertx);
    for (var route : routes) route.apply(router);
    return router;
  }
}

final class HttpSingle implements HttpRequest {
  private static final Logger log = getLogger(HttpSingle.class);

  private final Vertx vertx;
  private final HttpRoute route;

  HttpSingle(Vertx vertx, HttpRoute route) {
    this.vertx = vertx;
    this.route = route;
  }

  @Override
  public final Handler<HttpServerRequest> get() {
    return route.apply(router(vertx));
  }
}

File: src/main/java/io/trydent/olimpo/http/HttpRequest.kt
package io.trydent.olimpo.http

import io.vertx.core.Handler
import io.vertx.core.Vertx
import io.vertx.core.http.HttpServerRequest
import io.vertx.ext.web.Router.router
import org.slf4j.LoggerFactory.getLogger

interface HttpRequest : () -> Handler<HttpServerRequest> {
  companion object {
    fun routeSwitch(vertx: Vertx, vararg routes: HttpRoute): HttpRequest = HttpRouteSwitch(vertx, *routes)

    fun single(vertx: Vertx, route: HttpRoute): HttpRequest = HttpSingle(vertx, route)
  }
}

private fun Vertx.createRouter() = router(this)

internal class HttpRouteSwitch(private val vertx: Vertx, private vararg val switchRoutes: HttpRoute) : HttpRequest {
  private val log = getLogger(javaClass)

  override fun invoke(): Handler<HttpServerRequest> = vertx.createRouter().apply {
    switchRoutes.forEach { route -> route(this) }
  }
}

internal class HttpSingle(private val vertx: Vertx, private val route: HttpRoute) : HttpRequest {
  private val log = getLogger(javaClass)

  override fun invoke(): Handler<HttpServerRequest> = vertx.createRouter().apply {
    route(this).errorHandler(500) { log.error(it.failure().message) }
  }
}
